### DISCUSSION TITLE
Need some help on `postUpgradeTasks` config

### DISCUSSION THREAD

**Discussion by Justineo** [renovatebot/renovate#13102] [Request Help]
I'm using the Renovate GitHub Action with a monorepo managed by Lerna.

But Lerna has a bug when working with npm@7+ (https://github.com/lerna/lerna/issues/2845) that all indentation spaces in `package-lock.json` will be replaced with tabs. In order to work around this I added a `prettier --write packages/**/package-lock.json` in the `postUpgradeTasks` (see [here](https://github.com/ecomfe/veui/blob/7b87defc61ffa9ae2981113201f2d6352701b97a/.github/renovate.json#L17-L21)) and it works the first time Renovate made a PR and we didn't get massive changes to the indentation. But for the next PR the formatter seems to have missed a file (see [here](https://github.com/ecomfe/veui/pull/952/files#diff-7253cdae8febf9a2e314eb7b41775a243c3efcd19fca4dbb25766b64f2bf4e47)) but other files in the same PR seems to be correctly formatted.

It feels very weird so I'm wondering is there any chance the command executed at a wrong time or the upgrade process may have some racing conditions. Any help would be appreciated.

**Comment by rarkins** [comment]
It's perhaps because Renovate hasn't written out the changed file? Or maybe because the version of the file which Renovate thinks is right overwrites the one prettier changes. It's weird that only one file is missed out though. Is it repeatable, and always the same file?

**Reply by Justineo** [reply]
I manually closed previous PR and tried twice more and yes it's repeatable & always the same file. I'm pretty sure the command itself is correct as it worked for the previous version. Really have no idea what happened...

**Reply by rarkins** [reply]
If the branch remains when closing the PR then it's not a valid test because the branch will be reused. The commit needs to be regenerated

**Reply by Justineo** [reply]
Oh, I wasn't aware of that. I'll delete the branch and try again.

**Reply by Justineo** [reply]
![image](https://user-images.githubusercontent.com/1726061/146126876-371d3a9b-35bd-4afe-9d72-10ea33efe9aa.png)

Okay the branch is re-created and it turns out that the result is stable...

**Reply by Justineo** [reply]
I found the difference between the missed lock file and others. I have 2 packages relying on the same `dls-icons-vue` and the missed one updated to the latest version already. It seems that the lock file is changed by lerna itself but not considered as upgraded by Renovate so maybe the `postUpgradeTasks` won't run if there's "no upgrade" from Renovate's perspective?

**Reply by rarkins** [reply]
That's correct, postUpgradeTasks only run if there's an upgrade to begin with

**Reply by Justineo** [reply]
But Renovate did change the lock file though there was no upgrade involved. So does it make sense to add similar hook which is executed before Renovate is ready to commit the changes it made?

**Reply by rarkins** [reply]
It's hard to tell what's going on without a reproduction we can debug

**Reply by Justineo** [reply]
@rarkins I've created a minimal reproduction here: https://github.com/Justineo/lerna-renovate-repro

I created a monorepo with Lerna and there are two packages:

```
root
└─ packages
   ├─ my-foo → nanoid@3.1.30
   └─ my-bar → nanoid@3.1.29
```

The PR created by Renovate [here](https://github.com/Justineo/lerna-renovate-repro/pull/1) shows that:

* For `my-foo`, `nanoid` is already up-to-date and all spaces are replaced by tabs (by Lerna, and Prettier command isn't executed).
* For `my-bar`, `nanoid` is upgraded and Prettier command is executed as expected.

**Reply by rarkins** [reply]
Can you add a readme to that repo which describes (a) expected behavior, and (b) actual behavior?

**Reply by rarkins** [reply]
Please also document the config.js global config you used

**Reply by Justineo** [reply]
Sure, I've added a readme to the repro.

**Reply by rarkins** [reply]
The two packages have different values in `package.json`. Do you think that's partly causing it? i.e. one gets the `package.json` updated while one does not.

**Reply by Justineo** [reply]
Yes I believe that's what caused the difference. `my-foo` doesn't got updated but Lerna has already made changes to `package-lock.json`. `my-bar` is updated so the Prettier command seems to be executed as expected.

**Reply by Justineo** [reply]
Maybe we should only commit changes for files that got updated for deps version, and discard changes for those whose deps are not updated.
